### ISSUE TITLE
How to deal with db updates from the outside?

### ISSUE THREAD

**Issue by jreinert** [worona/meteorflux#25]
```
AppState.modify('tasks', (action, state = []) => {
	switch(action.type) {
		case 'TASK_SUBMITTED':
		case 'TASKS_SUBSCRIPTION_READY':
			state = Tasks.find();
			break;
	}
	return state;
});

AppState.modify('task', (action, state = {}) => {
	switch(action.type) {
		case 'TASK_SELECTED':
			state = Tasks.findOne({ _id: action._id });
			break;
		case 'TASK_CHANGED':
			state = action.attributes;
			break;
		case 'TASK_SUBMITTED':
			if (state._id) {
				Tasks.update(state._id, { $set: _.omit(state, '_id') });
			} else {
				let id = Tasks.insert(state);
				state = Tasks.findOne({ _id: id });
			}
			break;
	}
	return state;
});

Dispatcher.register((action) => console.debug(action.type, _.omit(action, 'type')));
Meteor.startup(() => TaskActions.subscribe());
```

This store works fine with one user editing. If I add a new task from `meteor mongo` the views won't be updated. Honestly I don't know how to deal with this and it seems like quite the show stopper...

**Comment by LuisHerranz** [comment]
@jreinert you are absolutely right.

To solve the mess between Flux (Action driven) and Meteor (Reactive Data driven) we are experimenting with a new framework called **First, Then, Finally**. The idea is this:

- When an **Action** is dispatched, **First** callbacks are triggered before anything else. Use them to write to the database, for example. These functions will be called only once.
- After that, all the reactive functions are triggered. These are typical **State** functions using new values from the database or other reactive sources like router params or different **State** values. These functions will be called as many times as they need if they are invalidated not only by **Action** but by other reactive sources.
- Once all the reactivity has finished and **State** is stable again, the **Finally** callbacks are triggered. Use them when you need to do staff with the updated **State** which can't be triggered more than once, like for example analytics or logging.

You can read more about it here:
https://github.com/worona/meteorflux/tree/devel/packages/first-then-finally

Feel free to experiment but please take into consideration that it is still under heavy development. 

Not really sure if all this is right in your case but this is an idea about how to implement your logic.

```javascript
First(() => {
  if (Action.is('TASK_SUBMITTED')) {
     if (Action._id)
        Tasks.update(Action._id, { $set: _.omit(Action.payload(), '_id') });
      else
        Tasks.insert(state);
    }
});

// Use State.set instead of Then for convenience.
State.set('tasks', (state = []) => {
  return Tasks.find();
});

State.set('task', (state = {}) => {
   switch(Action.type()) {
        case 'TASK_SELECTED':
            return Tasks.findOne({ _id: Action._id });
        case 'TASK_CHANGED':
            return Action.attributes;
        default:
            return state;
  }
});

Finally(() => console.debug(Action.type(), Action.payload()) );

// other stuff...
Meteor.startup(() => TaskActions.subscribe());
```

We are working with it right now. Function names are going to change probably, even the framework name.

**Comment by luisherranz** [comment]
You can take a look at the new API here:
https://github.com/worona/meteorflux/

We have been using it for a while and we are happy with it so we don't see any upcoming changes in the near future. I will close this now.
